server/src/main/java/org/elasticsearch/common/metrics/ExponentialBucketHistogram.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.common.metrics;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.LongAdder;
+
+/**
+ * A histogram with a fixed number of buckets of exponentially increasing width.
+ * <p>
+ * The bucket boundaries are defined by increasing powers of two, e.g.
+ * <code>
+ *     (-&infin;, 1), [1, 2), [2, 4), [4, 8), ..., [2^({@link #BUCKET_COUNT}-2), &infin;)
+ * </code>
+ * There are {@link #BUCKET_COUNT} buckets.
+ */
+public class ExponentialBucketHistogram {
+
+    public static int[] getBucketUpperBounds() {
+        int[] bounds = new int[17];
+        for (int i = 0; i < bounds.length; i++) {
+            bounds[i] = 1 << i;
+        }
+        return bounds;
+    }
+
+    private static int getBucket(long observedValue) {
+        if (observedValue <= 0) {
+            return 0;
+        } else if (LAST_BUCKET_LOWER_BOUND <= observedValue) {
+            return BUCKET_COUNT - 1;
+        } else {
+            return Long.SIZE - Long.numberOfLeadingZeros(observedValue);
+        }
+    }
+
+    // We add an extra bucket to track values greater (up Long.MAX_VALUE) than the defined bucket bounds
+    public static final int BUCKET_COUNT = getBucketUpperBounds().length + 1;
+
+    private static final long LAST_BUCKET_LOWER_BOUND = getBucketUpperBounds()[BUCKET_COUNT - 2];
+
+    private final LongAdder[] buckets;
+
+    public ExponentialBucketHistogram() {
+        buckets = new LongAdder[BUCKET_COUNT];
+        for (int i = 0; i < BUCKET_COUNT; i++) {
+            buckets[i] = new LongAdder();
+        }
+    }
+
+    public void addObservation(long observedValue) {
+        buckets[getBucket(observedValue)].increment();
+    }
+
+    /**
+     * @return An array of frequencies of handling times in buckets with upper bounds as returned by {@link #getBucketUpperBounds()}, plus
+     *         an extra bucket for handling times longer than the longest upper bound.
+     */
+    public long[] getSnapshot() {
```

Comment:
I changed this from `getHistogram` to `getSnapshot` because `histogram.getHistogram()` would look weird.